### ISSUE TITLE
RedundantConditionGivenDocblockType makes runtime validation of wildcard constants difficult (impossible?)

### ISSUE THREAD

**Issue by flaviovs** [vimeo/psalm#6214]
https://psalm.dev/r/e583127593

The code in the snippet is very common when doing validation of data (often coming from the user), so IMHO Psalm should not complain in cases like this.

Note: the issue is aggravated by the fact that apparently there's no way to locally suppress the message. Adding `/** @psalm-suppress RedundantConditionGivenDocblockType */` above line 13 does **not** supress the message (not to say it does not look right, since the suppression is inside a unrelated code block).

**Comment by weirdan** [comment]
Why don't you suppress it globally then?

**Comment by flaviovs** [comment]
Because this will suppress all other legitimate warnings about redundant conditions Psalm is supposed to catch, like the one in the warning documentation:

```php
<?php
/**
 * @param string $s
 *
 * @return void
 */
function foo($s) {
    if (is_string($s)) {};
}
```

**Comment by weirdan** [comment]
So you're trusting callers will pass strings where `string` is specified in the docblock, but don't think they will respect `Foo::BAR_*`?

**Comment by flaviovs** [comment]
The string example is just to illustrate that disabling the option globally is not a good workaround.

Notice that this issue is not about **trusting runtime input types**, but about **being able to check input data contents**, and the snippet I sent above is a legitimate, secure, and I'd say very common way of doing that.

Currently, the only way of getting rid of the warning  I could figure out (besides doing it globally, or making the code more complex than it needs to be) is to remove the last condition and use just an `else`. This makes sense from a static analysis point of view, where there is only one possible value left, however at runtime that can lead to bugs and security issues that are hard to find. This is specially true if you use Psalm in a library, and library users do not use it.

A practical example:

```php
<?php
class Foo {
    public const SEND_EMAIL = 'email';
    public const SEND_SMS = 'sms';
}

/** @param Foo::SEND_* $send_type */
function send_email_or_sms(string $send_type): void
{
    if ($send_type === Foo::SEND_EMAIL) {
        echo 'sending email...';
        // ...
    } elseif ($send_type === Foo::SEND_SMS) {
        echo 'sending sms...';
        // ...
    } else {
        die('Sorry. send_email_or_sms() can only send email or sms');
    }
}
```

However:

     ERROR: RedundantConditionGivenDocblockType - 14:15 - Docblock-defined type "sms" for $send_type is always string(sms)

Now we change the code to remove the warning:

```php
<?php

class Foo {
    public const SEND_EMAIL = 'email';
    public const SEND_SMS = 'sms';
}

/** @param Foo::SEND_* $send_type */
function send_email_or_sms(string $send_type): void
{
    if ($send_type === Foo::SEND_EMAIL) {
        echo 'sending email...';
        // ...
    } else {
        // According to Psalm, this can only be Foo::SEND_SMS, so we're good with just an "else".
        echo 'sending sms...';
        // ...
    }
}
```

Now:

```php
/** @var Foo::SEND_* */
$send_type = $conn->prepare("SELECT send_type FROM user WHERE user_id = ?")->execute([$user_id])->fetchColumn();
send_email_or_sms($send_type);
```

or (on a client app, one that perhaps does not use Psalm, or has *RedundantConditionGivenDocblockType* turned off):

```php
send_email_or_sms($_GET['send_type']);
```

Another nasty side-effect of the catch-all `else` is that if you add new constants (say `Foo::SEND_WHATSAPP`) then all code using `else` will happily process it as the "only option left", which, again, can lead to very hard to find bugs.

Lastly, one can replace `if (is constant1) ... elseif (is constant2) ... else { die() }` with a `switch` statement. **Psalm does not issue a warning in this case**, however there's a subtle semantic change in that `switch` does loose comparison, and this can lead to other bugs and security issues.
